File: io/lucamienert/basic/instruction/Instruction.java
package io.lucamienert.basic.instruction;

import java.util.Objects;
import java.util.Optional;

public class Instruction {

    private final OpCode opcode;
    private final Optional<string> stringOperand;
    private final Optional<string> intOperand;

    public Instruction(OpCode opcode) {
        this.opcode = opcode;
        this.stringOperand = Optional.empty();
        this.intOperand = Optional.empty();
    }

    public Instruction(OpCode opcode, String operand) {
        this.opcode = opcode;
        this.stringOperand = Optional.of(operand);
        this.intOperand = Optional.empty();
    }

    public Instruction(OpCode opcode, int operand) {
        this.opcode = opcode;
        this.stringOperand = Optional.empty();
        this.intOperand = Optional.of(operand);
    }

    public Opcode getOpcode() {
        return opcode;
    }

    public Optional<String> getStringOperand() {
        return stringOperand;
    }

    public Optional<Integer> getIntOperand() {
        return intOperand;
    }
}

File: src/BasicCompiler/Instructions/Instruction.cs
using BasicCompiler.Helper;

namespace BasicCompiler.Instructions;

public class Instruction
{
    public OpCode OpCode { get; set; }
    public Optional<string> StringOperand { get; set; }
    public Optional<int> IntOperand { get; set; }

    public Instruction(OpCode opCode)
    {
        OpCode = opCode;
        StringOperand = Optional<string>.Empty();
        IntOperand = Optional<int>.Empty();
    }

    public Instruction(OpCode opCode, string operand)
    {
        OpCode = opCode;
        StringOperand = Optional<string>.Of(operand);
        IntOperand = Optional<int>.Empty();
    }

    public Instruction(OpCode opCode, int operand)
    {
        OpCode = opCode;
        StringOperand = Optional<string>.Empty();
        IntOperand = Optional<int>.Of(operand);
    }
}
